File: backend/src/repository/postgres_user_repository.rs
use async_trait::async_trait;
use sqlx::PgPool;
use uuid::Uuid;

use crate::domain::entity::user::User;
use crate::domain::valueobject::email::Email;
use crate::domain::valueobject::username::Username;
use crate::repository::user_repository::{UserRepository, UserRepositoryError};

/// PostgreSQLユーザーリポジトリ
pub struct PostgresUserRepository {
    pool: PgPool,
}

impl PostgresUserRepository {
    /// 新しいPostgreSQLリポジトリを作成
    #[must_use]
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    /// データベースマイグレーションを実行
    ///
    /// # Errors
    ///
    /// - マイグレーション実行に失敗した場合
    pub async fn migrate(&self) -> Result<(), UserRepositoryError> {
        sqlx::query(
            r#"
            CREATE TABLE IF NOT EXISTS users (
                user_id UUID PRIMARY KEY,
                username VARCHAR(20) NOT NULL UNIQUE,
                email VARCHAR(254) NOT NULL UNIQUE,
                password_hash VARCHAR(60) NOT NULL,
                created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
            )
            "#,
        )
        .execute(&self.pool)
        .await
        .map_err(|e| UserRepositoryError::DatabaseError(e.to_string()))?;

        Ok(())
    }
}

#[async_trait]
impl UserRepository for PostgresUserRepository {
    async fn save(&self, user: &User) -> Result<(), UserRepositoryError> {
        sqlx::query(
            r#"
            INSERT INTO users (user_id, username, email, password_hash)
            VALUES ($1, $2, $3, $4)
            ON CONFLICT (user_id) DO UPDATE
            SET username = $2, email = $3, password_hash = $4, updated_at = NOW()
            "#,
        )
        .bind(user.user_id())
        .bind(user.username().as_str())
        .bind(user.email().as_str())
        .bind(user.password_hash())
        .execute(&self.pool)
        .await
        .map_err(|e| UserRepositoryError::DatabaseError(e.to_string()))?;

        Ok(())
    }

    async fn find_by_id(&self, user_id: &Uuid) -> Result<Option<User>, UserRepositoryError> {
        let row = sqlx::query_as::<_, (Uuid, String, String, String)>(
            r#"
            SELECT user_id, username, email, password_hash
            FROM users
            WHERE user_id = $1
            "#,
        )
        .bind(user_id)
        .fetch_optional(&self.pool)
        .await
        .map_err(|e| UserRepositoryError::DatabaseError(e.to_string()))?;

        match row {
            Some((user_id, username, email, password_hash)) => {
                let username = Username::new(&username).map_err(|e| {
                    UserRepositoryError::DatabaseError(format!("Invalid username: {e}"))
                })?;
                let email = Email::new(&email).map_err(|e| {
                    UserRepositoryError::DatabaseError(format!("Invalid email: {e}"))
                })?;
                let user = User::from_repository(user_id, username, email, &password_hash)
                    .map_err(|e| {
                        UserRepositoryError::DatabaseError(format!("Invalid password hash: {e}"))
                    })?;
                Ok(Some(user))
            }
            None => Ok(None),
        }
    }

    async fn find_by_email(&self, email: &Email) -> Result<Option<User>, UserRepositoryError> {
        let row = sqlx::query_as::<_, (Uuid, String, String, String)>(
            r#"
            SELECT user_id, username, email, password_hash
            FROM users
            WHERE email = $1
            "#,
        )
        .bind(email.as_str())
        .fetch_optional(&self.pool)
        .await
        .map_err(|e| UserRepositoryError::DatabaseError(e.to_string()))?;

        match row {
            Some((user_id, username, email, password_hash)) => {
                let username = Username::new(&username).map_err(|e| {
                    UserRepositoryError::DatabaseError(format!("Invalid username: {e}"))
                })?;
                let email = Email::new(&email).map_err(|e| {
                    UserRepositoryError::DatabaseError(format!("Invalid email: {e}"))
                })?;
                let user = User::from_repository(user_id, username, email, &password_hash)
                    .map_err(|e| {
                        UserRepositoryError::DatabaseError(format!("Invalid password hash: {e}"))
                    })?;
                Ok(Some(user))
            }
            None => Ok(None),
        }
    }

    async fn find_by_username(
        &self,
        username: &Username,
    ) -> Result<Option<User>, UserRepositoryError> {
        let row = sqlx::query_as::<_, (Uuid, String, String, String)>(
            r#"
            SELECT user_id, username, email, password_hash
            FROM users
            WHERE username = $1
            "#,
        )
        .bind(username.as_str())
        .fetch_optional(&self.pool)
        .await
        .map_err(|e| UserRepositoryError::DatabaseError(e.to_string()))?;

        match row {
            Some((user_id, username, email, password_hash)) => {
                let username = Username::new(&username).map_err(|e| {
                    UserRepositoryError::DatabaseError(format!("Invalid username: {e}"))
                })?;
                let email = Email::new(&email).map_err(|e| {
                    UserRepositoryError::DatabaseError(format!("Invalid email: {e}"))
                })?;
                let user = User::from_repository(user_id, username, email, &password_hash)
                    .map_err(|e| {
                        UserRepositoryError::DatabaseError(format!("Invalid password hash: {e}"))
                    })?;
                Ok(Some(user))
            }
            None => Ok(None),
        }
    }

    async fn exists_by_email(&self, email: &Email) -> Result<bool, UserRepositoryError> {
        let exists = sqlx::query_scalar::<_, bool>(
            r#"
            SELECT EXISTS(SELECT 1 FROM users WHERE email = $1)
            "#,
        )
        .bind(email.as_str())
        .fetch_one(&self.pool)
        .await
        .map_err(|e| UserRepositoryError::DatabaseError(e.to_string()))?;

        Ok(exists)
    }

    async fn exists_by_username(&self, username: &Username) -> Result<bool, UserRepositoryError> {
        let exists = sqlx::query_scalar::<_, bool>(
            r#"
            SELECT EXISTS(SELECT 1 FROM users WHERE username = $1)
            "#,
        )
        .bind(username.as_str())
        .fetch_one(&self.pool)
        .await
        .map_err(|e| UserRepositoryError::DatabaseError(e.to_string()))?;

        Ok(exists)
    }
}

File: backend-hono/src/repository/postgres/PostgresUserRepository.ts
import { eq } from 'drizzle-orm';
import { DbType } from '../../db/connection';
import { users } from '../../db/schema';
import { User } from '../../domain/entity/User';
import { UserRepository, UserRepositoryError } from '../interface/UserRepository';

export class PostgresUserRepository implements UserRepository {
  constructor(private readonly db: DbType) {}

  async save(user: User): Promise<void> {
    try {
      await this.db.insert(users).values({
        userId: user.getUserId(),
        username: user.getUsername(),
        email: user.getEmail(),
        passwordHash: user.getPasswordHash(),
      });
    } catch (error) {
      throw new UserRepositoryError(`Failed to save user: ${error}`);
    }
  }

  async findById(userId: string): Promise<User | null> {
    try {
      const result = await this.db.select().from(users).where(eq(users.userId, userId)).limit(1);

      if (result.length === 0) {
        return null;
      }

      const row = result[0];
      return User.fromRepository(row.userId, row.username, row.email, row.passwordHash);
    } catch (error) {
      throw new UserRepositoryError(`Failed to find user by ID: ${error}`);
    }
  }

  async findByEmail(email: string): Promise<User | null> {
    try {
      const result = await this.db
        .select()
        .from(users)
        .where(eq(users.email, email.toLowerCase()))
        .limit(1);

      if (result.length === 0) {
        return null;
      }

      const row = result[0];
      return User.fromRepository(row.userId, row.username, row.email, row.passwordHash);
    } catch (error) {
      throw new UserRepositoryError(`Failed to find user by email: ${error}`);
    }
  }

  async findByUsername(username: string): Promise<User | null> {
    try {
      const result = await this.db
        .select()
        .from(users)
        .where(eq(users.username, username))
        .limit(1);

      if (result.length === 0) {
        return null;
      }

      const row = result[0];
      return User.fromRepository(row.userId, row.username, row.email, row.passwordHash);
    } catch (error) {
      throw new UserRepositoryError(`Failed to find user by username: ${error}`);
    }
  }

  async existsByEmail(email: string): Promise<boolean> {
    try {
      const result = await this.db
        .select({ userId: users.userId })
        .from(users)
        .where(eq(users.email, email.toLowerCase()))
        .limit(1);

      return result.length > 0;
    } catch (error) {
      throw new UserRepositoryError(`Failed to check email existence: ${error}`);
    }
  }

  async existsByUsername(username: string): Promise<boolean> {
    try {
      const result = await this.db
        .select({ userId: users.userId })
        .from(users)
        .where(eq(users.username, username))
        .limit(1);

      return result.length > 0;
    } catch (error) {
      throw new UserRepositoryError(`Failed to check username existence: ${error}`);
    }
  }
}
